st2api/st2api/controllers/exp/inquiries.py

```diff
@@ -0,0 +1,295 @@
+            raw_inquiry = raw_inquiry.__json__()
+
+        desired_fields = ["tag", "ttl", "users", "roles", "schema"]
+
+        inquiry = {
+            "id": raw_inquiry["id"]
+        }
+        for field in desired_fields:
+            inquiry[field] = raw_inquiry["result"][field]
+
+        return inquiry
+
+    def _mark_inquiry_complete(self, inquiry_id, result):
+        """Mark Inquiry as completed
+
+        This function updates the local LiveAction and Execution with a successful
+        status as well as call the "post_run" function for the Inquirer runner so that
+        the appropriate callback function is executed
+
+        :param inquiry: The Inquiry for which the response is given
+        :param requester_user: The user providing the response
+
+        :rtype: bool - True if requester_user is able to respond. False if not.
+        """
+
+        # Update inquiry's execution result with a successful status and the validated response
+        liveaction_db = action_utils.update_liveaction_status(
+            status=action_constants.LIVEACTION_STATUS_SUCCEEDED,
+            runner_info=system_info.get_process_info(),
+            result=result,
+            liveaction_id=inquiry_id)
+        executions.update_execution(liveaction_db)
+
+        # Call Inquiry runner's post_run to trigger callback to workflow
+        runner_container = get_runner_container()
+        action_db = get_action_by_ref(liveaction_db.action)
+        runnertype_db = get_runnertype_by_name(action_db.runner_type['name'])
+        runner = runner_container._get_runner(runnertype_db, action_db, liveaction_db)
+        runner.post_run(status=action_constants.LIVEACTION_STATUS_SUCCEEDED, result=result)
+
+        return liveaction_db
+
+    def _can_respond(self, inquiry, requester_user):
+        """Determine if requester_user is permitted to respond based on parameters
+
+        This determines if the requesting user has permission to respond to THIS inquiry.
+        Note this is NOT RBAC, and you should still protect the API endpoint with RBAC
+        where appropriate.
+
+        :param inquiry: The Inquiry for which the response is given
+        :param requester_user: The user providing the response
+
+        :rtype: bool - True if requester_user is able to respond. False if not.
+        """
+
+        # Deny by default
+        roles_passed = False
+        users_passed = False
+
+        # Determine role-level permissions
+        roles = inquiry.get('roles')
+        if roles:
+            for role in roles:
+                LOG.debug("Checking user %s is in role %s" % (requester_user, role))
+                LOG.debug(rbac_utils.user_has_role(requester_user, role))
+                # TODO(mierdin): Note that this will always return True if Rbac is not enabled
+                # Need to test with rbac enabled and configured
+                if rbac_utils.user_has_role(requester_user, role):
```

Comment:
OK, so it looks like the behavior is "OR".

Is there a reason when we would want "AND" behavior instead?

(just thinking out loud)

Reply:
If you're talking about the roles list, yes, in the current approach, you would only need to be a member of one of the listed roles. I don't **think** there's a solid use case for specifying that the user must be in ALL of the listed roles but am open to hearing them. For now, the approach is OR (with both users and roles)